doc/man3/SSL_read_early.pod

```diff
@@ -0,0 +1,209 @@
+ uint32_t SSL_get_max_early_data(const SSL_CTX *s);
+ uint32_t SSL_SESSION_get_max_early_data(const SSL_SESSION *s);
+
+ int SSL_write_early(SSL *s, const void *buf, size_t num, size_t *written);
+
+ int SSL_read_early(SSL *s, void *buf, size_t num, size_t *readbytes);
+
+ int SSL_get_early_data_status(const SSL *s);
+
+=head1 DESCRIPTION
+
+These functions are used to send and recieve early data. Early data can be sent
+by the client immediately after its initial ClientHello without having to wait
+for the server to complete the handshake. Early data can only be sent if a
+session has previously been established with the server, and the server is known
+to support it.
+
+Early data has weaker security properties than other data sent over an SSL/TLS
+connection. In particular the data is not forward secret and the server has no
+guarantees that the same early data was not replayed across multiple
+connections. For this reason extreme care should be exercised when using early
+data.
+
+When a server receives early data it may opt to immediately respond by sending
+application data back to the client. Data sent by the server at this stage is
+done before the full handshake has been completed. Specifically the client's
+authentication messages have not yet been received, i.e. the client is
+unauthenticated at this point.
+
+A server or client can determine whether the full handshake has been completed
+or not by calling L<SSL_is_init_finished(3)>.
+
+[[TODO(TLS1.3): The server uses SSL_write_ex()/SSL_write() to send data to an
+unauthenticated client. Should we create a separate function for this to avoid
+accidents??]]
+
+On the client side the function SSL_SESSION_get_max_early_data() can be used to
+determine whether a session established with a server can be used to send early
+data. If the session cannot be used then this function will return 0. Otherwise
+it will return the maximum number of early data bytes that can be sent.
+
+A client uses the function SSL_write_early() to send early data. This function
+works in the same way as the L<SSL_write_ex(3)> function, but with a few
+differences. Refer to the L<SSL_write_ex(3)> documentation for
+information on how to write bytes to the underlying connection, and how to
+handle any errors that may arise. This page will detail the differences between
+SSL_write_early() and L<SSL_write_ex(3)>.
+
+SSL_write_early() must be the first IO function called on a new connection, i.e.
+it must occur before any calls to L<SSL_write_ex(3)>, L<SSL_read_ex(3)>,
+L<SSL_connect(3)>, L<SSL_do_handshake(3)> or other similar functions. It may be
+called multiple times to stream data to the server, but the total number of
+bytes written must not exceed the value returned from
+SSL_SESSION_get_max_early_data(). Once the initial SSL_write_early() call has
+completed successfully the client may interleave calls to L<SSL_read_ex(3)> and
+L<SSL_read(3)> with calls to SSL_write_early() as required.
```

Comment:
Well, here the client is communicating with a fully authenticated server - its not going to get any more messages from the server. Actually even after the client finishes its side of the handshake it is not going to know whether the records it is reading were sent by the server before or after it authenticated the client. That's just the way the protocol works. So I think SSL_read/SSL_read_ex is correct here.